<a href="https://colab.research.google.com/github/kla55/langchain-llm/blob/main/Langchain_RAG_app_example.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
pip install faiss-cpu transformers sentence-transformers

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 27.5/27.5 MB 13.8 MB/s eta 0:00:00


In [2]:
# Sample dataset (can be replaced with any open-source text data)
data = [
    {"id": 1, "text": "What is COVID-19?", "answer": "COVID-19 is caused by the SARS-CoV-2 virus."},
    {"id": 2, "text": "What are symptoms of COVID-19?", "answer": "Symptoms include fever, cough, and fatigue."},
    {"id": 3, "text": "How does COVID-19 spread?", "answer": "COVID-19 spreads primarily through respiratory droplets."},
]

In [3]:
from sentence_transformers import SentenceTransformer
import faiss
import numpy as np



/usr/local/lib/python3.10/dist-packages/huggingface_hub/utils/_auth.py:94: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


modules.json:   0%|          | 0.00/349 [00:00<?, ?B/s]

config_sentence_transformers.json:   0%|          | 0.00/116 [00:00<?, ?B/s]

README.md:   0%|          | 0.00/10.7k [00:00<?, ?B/s]

sentence_bert_config.json:   0%|          | 0.00/53.0 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/612 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/90.9M [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/350 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/232k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/466k [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/112 [00:00<?, ?B/s]

1_Pooling/config.json:   0%|          | 0.00/190 [00:00<?, ?B/s]

In [10]:
embedding_model = SentenceTransformer('all-MiniLM-L6-v2')
text = [entry['text'] for entry in data]

# Create embeddings for the dataset
embeddings = embedding_model.encode(text)

# Build a FAISS index
dimensions = embeddings.shape[1]
index = faiss.IndexFlatL2(dimensions) # L2 similarity
index.add(np.array(embeddings)) # Add embeddings to the index

In [9]:
dimensions

384

# Build a FAISS index
**Embedding**\
The embedding is a 2D NumPy array or similar structure where each row corresponds to the vector representation (embedding) of a text/document, and each column represents a dimension of that vector. \
Example: If embeddings has a shape (3, 384), it means there are 3 embeddings, each with 384 dimensions.\
Purpose: dimension stores the number of dimensions per embedding (384 in this example). FAISS needs this to initialize the index properly.

**IndexFlatL2()**\
FAISS (Facebook AI Similarity Search) provides a way to perform fast similarity searches over high-dimensional vectors.\

**IndexFlatL2:**\
This creates a "flat" (non-hierarchical) index that computes similarities using L2 (Euclidean) distance.\
L2 distance is defined as the square root of the sum of squared differences between corresponding elements in two vectors: L2 distance = ∑sqrt((𝑥𝑖−𝑦𝑖)2)\
Purpose: The IndexFlatL2 object is initialized to accept embeddings of the specified dimension and perform searches based on Euclidean distance.

**np.array(embeddings):**\
Converts the embeddings into a NumPy array if it isn't already.
This ensures compatibility with FAISS, which expects a NumPy array for indexing.

**index.add():**\
Adds the embeddings to the FAISS index. Each row in embeddings becomes a searchable vector in the index.\
Purpose: The index now contains the embeddings and is ready to perform similarity searches.

In [7]:
dimension = embeddings.shape[1]
index = faiss.IndexFlatL2(dimension)  # L2 similarity
index.add(np.array(embeddings))  # Add embeddings to the index

array([[-0.04122965,  0.05281242, -0.0756428 , ..., -0.03431396,
         0.08876473,  0.02983131],
       [ 0.02599608,  0.00603665, -0.0270176 , ..., -0.06681953,
        -0.03288478,  0.04134002],
       [ 0.00266531, -0.00160182, -0.03908974, ..., -0.03787267,
         0.03340581,  0.08272062]], dtype=float32)

In [ ]:
def retrieve_relevant_text(query, index, embedding_model, texts, k=1):
    query_embedding = embedding_model.encode([query])
    distances, indices = index.search(query_embedding, k)
    return [texts[i] for i in indices[0]]

# Example query
query = "What causes COVID-19?"
retrieved_texts = retrieve_relevant_text(query, index, embedding_model, texts)
print("Retrieved Text:", retrieved_texts)